source/tutorials/module-system/module-system.md

```diff
@@ -0,0 +1,1289 @@
+# marker.nix
+-  userType = lib.types.submodule {
++  userType = lib.types.submodule ({ name, ... }: {
+     options = {
+       departure = lib.mkOption {
+         type = markerType;
+         default = {};
+       };
+     };
+-  };
+```
+
+In this case, you don't easily have access to the name from the marker submodules `label` option, where you otherwise could set a `default` value.
+
+Instead you can use the `config` section of the `user` submodule to set a default, like so:
+
+```diff
+# marker.nix
++
++    config = {
++      departure.style.label = lib.mkDefault
++        (firstUpperAlnum name);
++    };
++  });
+
+ in {
+
+```
+
+:::{note}
+Module options have a *priority*, represented as an integer, which determines the precedence for setting the option to a particular value.
+When merging values, the priority with lowest numeric value wins.
+
+The `lib.mkDefault` modifier sets the priority of its argument value to 1000, the lowest precedence.
+
+This ensures that other values set for the same option will prevail.
+:::
+
+## Marker Styling: Color
+
+For better visual contrast, it would be helpful to have a way to change the *color* of a marker.
+
+Here you will use two new type-functions for this:
+- `either <this> <that>`, which takes two types as arguments, and allows either of them
+- `enum [ <allowed values> ]`, which takes a list of allowed values, and allows any of them
+
+In the `let` block, add the following `colorType` option, which can hold strings containing either some given color names or an RGB value add the new compound type:
+
+```diff
+# marker.nix
+     ...
+     (builtins.match "[^A-Z0-9]*([A-Z0-9]).*"
+     (lib.toUpper str));
+
++  # Either a color name or `0xRRGGBB`
++  colorType = lib.types.either
++    (lib.types.strMatching "0x[0-9A-F]{6}")
++    (lib.types.enum [
++      "black" "brown" "green" "purple" "yellow"
++      "blue" "gray" "orange" "red" "white" ]);
++
+   markerType = lib.types.submodule {
+     options = {
+       location = lib.mkOption {
+```
+
+This allows either strings that matche a 24-bit hexadecimal number or are equal to one of the specified color names.
+
+At the bottom of the `let` block, add the `style.color` option and specify a default value:
+
+```diff
+# marker.nix
+           (lib.types.strMatching "[A-Z0-9]");
+         default = null;
+       };
++
++      style.color = lib.mkOption {
++        type = colorType;
++        default = "red";
++      };
+     };
+   };
+```
+
+Now add an entry to the `paramForMarker` list which makes use of the new option:
+
+```diff
+# marker.nix
+               (marker.style.label != null)
+               "label:${marker.style.label}"
+             ++ [
++              "color:${marker.style.color}"
+               "$(geocode ${
+                 lib.escapeShellArg marker.location
+               })"
+```
+
+## Marker Styling: Size
+
+In case you set many different markers, it would be helpful to have the ability to change their size individually.
```

Comment:
Because this introduces `enum` by itself, I would move this up, before Color, which introduces `enum` inside an `either`